### PR TITLE
Use curl-multi

### PR DESCRIPTION
**Pull request by Learath2**
Supersedes #5092, I am still not very sure about this structure either, but I think it's time someone else takes a look.

Remaining concerns:
- [ ] Should errors in the http module just be fatal?
- [ ] Is this really better than #5092's approach of turning http into a kernel interface?
- [ ] Changes in serverbrowser and register

I think @heinrich5991 might be able to come up with better fixes for the serverbrowser and the register component.

## Checklist

- [x] Tested the change ingame
- [ ] Provided screenshots if it is a visual change
- [ ] Tested in combination with possibly related configuration options
- [ ] Written a unit test (especially base/) or added coverage to integration test
- [x] Considered possible null pointers and out of bounds array indexing
- [ ] Changed no physics that affect existing maps
- [x] Tested the change with [ASan+UBSan or valgrind's memcheck](https://github.com/ddnet/ddnet/#using-addresssanitizer--undefinedbehavioursanitizer-or-valgrinds-memcheck) (optional)

### CODE REVIEW

**Review by heinrich5991 — COMMENTED** [review]
These are my thoughts. Everything looks like an improvement even without these comments handled. What do you think about these? Don't have to immediately implement them before giving your thoughts.

I'm interested in why you split src/engine/shared/engine.cpp.

Would it be possible to let the game server run HTTPS requests for one more second after it quits?

The engine becomes pretty much a singleton due to how the job pools are handled. Perhaps the different runners could get a hardcoded ID instead that we manually keep unique?

The additions in src/base/system.h could get some documentation, plus maybe a test. The HttpInit does some testing at the very start, perhaps that could also be used? Since `net_loop` is only used on Windows, could it maybe just be `pipe2` on unixy systems?

**Comment by Learath2** [comment]
I split `src/engine/shared/engine.cpp` mostly because I wanted to see the interface separately, it was mostly a tool to help me think. It also makes it more uniform with the rest of the code.

It should be possible to keep it running an extra second, I could also make it wait out the remaining requests if that is the concern.

Hm, isn't the engine already a singleton, along with most everything else in this weird kernel structure? I did want unique hardcoded IDs instead but I didn't want to manually keep them unique. Maybe I can come up with some template magic or preprocessor trickery. Will think about it.

Things in `base/system` will definitely get documentation. I was wondering if I should wrap `pipe2`, `write` and `read` as well since we tend to avoid platform specific includes inside the codebase. You mean `net_loop` should be `pipe2`? It's not really the same thing though, we need a better name for it since windows uses a self connected socket but unix uses a pipe with two distinct sides.

**Comment by heinrich5991** [comment]
> I split src/engine/shared/engine.cpp mostly because I wanted to see the interface separately, it was mostly a tool to help me think. It also makes it more uniform with the rest of the code.

I see. I hope no one starts including the new engine.h.

> It should be possible to keep it running an extra second, I could also make it wait out the remaining requests if that is the concern.

The extra second is for the self-deletion of the masterserver entries. This is just best-effort, it shouldn't delay the shutdown indefinitely.

> Hm, isn't the engine already a singleton, along with most everything else in this weird kernel structure?

I think the engine is instantiated in the tests, maybe even more than once.

> I did want unique hardcoded IDs instead but I didn't want to manually keep them unique.

We could maybe put them into a single file, e.g. into engine.h where you then always have to add your new runners.

> You mean `net_loop` should be `pipe2`? It's not really the same thing though, we need a better name for it since windows uses a self connected socket but unix uses a pipe with two distinct sides.

Yes, needs a better name. "selectable file descriptor supporting self-writes". :P I.e. I haven't come up with something good.

**Comment by Learath2** [comment]
@heinrich5991 couldn't figure out a way to merge them, so I made `io_pipe_*` and documented both.

File: src/engine/shared/http.cpp [hunk 1]
```diff
@@ -15,33 +15,275 @@
...
+	// print curl version
+	{
+		curl_version_info_data *pVersion = curl_version_info(CURLVERSION_NOW);
+		dbg_msg("http", "libcurl version %s (compiled = " LIBCURL_VERSION ")", pVersion->version);
+	}
+
+	m_State = RUNNING;
+	m_Cv.notify_one();
+	InitL.unlock();
+	InitL.release();
```
Comment (edg-l) [inline src/engine/shared/http.cpp:134]: why unlock and release?

**Comment by heinrich5991** [comment]
> I think the engine is instantiated in the tests, maybe even more than once.

Apparently the engine is not instantiated in the tests at all, I wonder why we even have `CreateTestEngine`… Perhaps make that explicit by adding a `dbg_assert` that no engine has been created before (e.g. by checking `IJob::m_sRunner`).

Otherwise, this looks good to me. I don't need to talk about it anymore, I think.

**Comment by heinrich5991** [comment]
Status: Needs rebase. I really want to have this PR. ^^

**Comment by heinrich5991** [comment]
Status: Needs rebase.

**Comment by heinrich5991** [comment]
Superseded by #7683 (merged already).

**Pull request closed without merging**